packages/@aws-cdk/aws-lambda/README.md

```diff
@@ -172,10 +173,76 @@ fn.addPermission('my-service Invocation', {
 });
 ```
 
-For more information, see [Resource-based
-policies](https://docs.aws.amazon.com/lambda/latest/dg/access-control-resource-based.html)
+You can also restrict permissions given to AWS services by providing
+a source account or ARN (representing the account and identifier of the resource
+that accesses the function or layer).
+
+For more information, see
+[Granting function access to AWS services](https://docs.aws.amazon.com/lambda/latest/dg/access-control-resource-based.html#permissions-resource-serviceinvoke)
 in the AWS Lambda Developer Guide.
 
+### Grant function access to an AWS Organization
+
+```ts
+// Grant permissions to an entire AWS organization
+declare const fn: lambda.Function;
+const orgPrincipal = new iam.OrganizationPrincipal('o-xxxxxxxxxx');
+
+fn.grantInvoke(orgPrincipal);
```

Comment:
The first order of business is probably figuring out whether this is an API we want to support.

I've worked very hard to make this possible, but I can be convinced that it isn't needed. Basically, my argument is that it is reasonable to configure permissions in this manner, and expect `OrganizationPrincipal` to be an object that can be passed into the `principal: iam.IPrincipal` property. The [docs](https://docs.aws.amazon.com/lambda/latest/dg/access-control-resource-based.html#permissions-resource-xorginvoke) shows an example that is exactly this use case -- granting permissions to an entire organization.

However, this isn't what lambda accepts and there's some fancy footwork in the backend to make this happen. Basically, I've translated this call into `principal: *` and `principalOrgID: o-xxxxxxxxxx`. That is what the equivalent call via aws CLI does.

I guess the question is, does all this work make sense in making a smoother API, or should we force our users to utilize the `principalOrg` property and confine them to one way of configuring this permission.

Reply:
I think my original thinking here was that it would look something like this:

```ts
fn.grantInvoke(new OrganizationPrincipal('o-1234'));

fn.grantInvoke(new OrganizationPrincipal('o-1234', new AccountPrincipal('111111111111')));
```

Now, I see that `OrganizationPrincipal` doesn't have a 2nd argument yet, but it could have that, and it could render the *regular* principal while `OrganizationPrincpal` adds a condition on top. We could add a convenience method on `BasePrincipal` so that an alternative constructor could be:

```ts
fn.grantInvoke(new AccountPrincipal('111111111111').inOrganization('o-1234'));
```

But that's sugar on top.

`principalOrg` will still need to be in `addPermission` though I think, but as a regular string (and we should not be documenting its use too much).

Reply:
See #20109. I love the `inOrganization` idea. I don't like adding a 2nd argument to `OrganizationPrincipal` as much. It doesn't make sense to me, because `new OrganizationPrincipal('o-1234', new AccountPrincipal('111111111111')` would not be an `OrganizationPrincipal` but an `AccountPrincipal`. I think we're better of exposing one nice way to do it, and steering users to that.